cli/enry/main.go

```diff
@@ -159,21 +167,94 @@ func printJson(out map[string][]string, buff *bytes.Buffer) {
 	buff.WriteByte('\n')
 }
 
-func printPercents(out map[string][]string, buff *bytes.Buffer) {
-	var fileCountList enry.FileCountList
-	total := 0
-	for name, language := range out {
-		fc := enry.FileCount{Name: name, Count: len(language)}
-		fileCountList = append(fileCountList, fc)
-		total += len(language)
+// filelistError represents a failed operation that took place across multiple files.
+type filelistError []string
+
+func (e filelistError) Error() string {
+	return fmt.Sprintf("Could not process the following files:\n%s", strings.Join(e, "\n"))
+}
+
+func printPercents(fSummary map[string][]string, buff *bytes.Buffer, mode string) {
+	// Select the way we quantify 'amount' of code.
+	var reducer func([]string) (float64, filelistError)
+	switch mode {
+	case "file":
+		reducer = fileCountValues
+	case "line":
+		reducer = lineCountValues
+	case "byte":
+		reducer = byteCountValues
+	default:
+		reducer = fileCountValues
 	}
-	// Sort the fileCountList in descending order of their count value.
-	sort.Sort(sort.Reverse(fileCountList))
 
-	for _, fc := range fileCountList {
-		percent := float32(fc.Count) / float32(total) * 100
-		buff.WriteString(fmt.Sprintf("%.2f%%	%s\n", percent, fc.Name))
+	// Reduce the list of files to a quantity of file type.
+	var total float64
+	fileValues := make(map[string]float64)
+	keys := []string{}
+	var unreadableFiles filelistError
+	for fType, files := range fSummary {
+		val, err := reducer(files)
+		if err != nil {
+			unreadableFiles = append(unreadableFiles, err...)
+		}
+		fileValues[fType] = val
+		keys = append(keys, fType)
+		total += val
 	}
+
+	// Slice the keys by their quantity (file count, line count, byte size, etc.).
+	sort.Slice(keys, func(i, j int) bool {
+		return fileValues[keys[i]] > fileValues[keys[j]]
+	})
+
+	// Calculate and write percentages of each file type.
+	for _, fType := range keys {
+		val := fileValues[fType]
+		percent := val / total * 100.0
+		buff.WriteString(fmt.Sprintf("%.2f%%\t%s\n", percent, fType))
+		if unreadableFiles != nil {
+			buff.WriteString(fmt.Sprintf("\n%s", unreadableFiles.Error()))
+		}
+	}
+}
+
+func fileCountValues(files []string) (float64, filelistError) {
+	return float64(len(files)), nil
+}
+
+func lineCountValues(files []string) (float64, filelistError) {
+	var filesErr filelistError
+	var t float64
+	for _, fName := range files {
+		content, err := ioutil.ReadFile(fName)
+		if err != nil {
+			filesErr = append(filesErr, fName)
+			continue
+		}
+		l, _ := getLines(content)
+		t += float64(l)
+	}
+	return t, filesErr
+}
+
+func byteCountValues(files []string) (float64, filelistError) {
+	var filesErr filelistError
+	var t float64
+	for _, fName := range files {
+		f, err := os.Open(fName)
```

Comment:
You are leaking file descriptors here. Close file directly after `Stat`, but before the `if` statement.